Add table text wrap toggle

LDSamson opened the pull request:
Mostly solves #224 .

Provides the use to enable text wrapping. The option is off by default, and a warning is shown that it might be slow to enable it for big tables

<img width="1689" height="820" alt="grafik" src="https://github.com/user-attachments/assets/f9ad9b06-7cd3-40fa-91d0-694d4144bd19" />

LDSamson committed:
Add toggle to enable multi-linel table rows in common forms

(cherry picked from commit 2d34a1d3eca92d059bb2ddd8bc6b53a2b41e465b)

LDSamson committed:
Update toggle text

LDSamson committed:
Also enable toggle in study forms

LDSamson committed:
Update inputid to enable_text_wrap for consistency

LDSamson committed:
Do not allow text wrap in form review mode

LDSamson committed:
Merge remote-tracking branch 'origin/dev' into ls_224_add_table_text_wrap_toggle

LDSamson committed:
Fix snaps for app feature 3, make them more focused

LDSamson committed:
Update news

LDSamson committed:
Bump version

LDSamson committed:
Capture table text wrap in feature 1 test

jthompson-arcus commented on R/mod_common_forms.R:
```diff
@@ -148,6 +169,7 @@ mod_common_forms_server <- function(
         form_items = form_items,
         active_subject = active_subject,
         show_all = reactive(isTRUE(input$show_all_data) | identical(session$userData$review_type(), "form") ),
+        enable_text_wrap = reactive(isTRUE(input$enable_text_wrap)),
```
```suggestion
        enable_text_wrap = reactive(isTRUE(input$enable_text_wrap) & identical(session$userData$review_type(), "subject")),
```
For consistency

jthompson-arcus commented on R/mod_study_forms.R:
```diff
@@ -176,6 +200,7 @@ mod_study_forms_server <- function(
       active_subject = active_subject,
       form_items = form_items,
       show_all = reactive(isTRUE(input$show_all) | identical(session$userData$review_type(), "form")), 
+      enable_text_wrap = reactive(isTRUE(input$enable_text_wrap)),
```
```suggestion
      enable_text_wrap = reactive(isTRUE(input$enable_text_wrap) & identical(session$userData$review_type(), "subject")),
```
For consistency

jthompson-arcus reviewed (changes requested):
Enabling text wrap does not appear to be working. For instance checkout subject BEL_04_772 and the Medication tab. There are 15 records. If you "enable text wrap" you can only view 10 of them.

LDSamson committed:
Update R/mod_study_forms.R

Co-authored-by: Jeff Thompson <160783290+jthompson-arcus@users.noreply.github.com>

LDSamson committed:
Update R/mod_common_forms.R

Co-authored-by: Jeff Thompson <160783290+jthompson-arcus@users.noreply.github.com>

jthompson-arcus commented:
I think generally I'm a bit confused by this PR. In the common events the text wrapping is disabled on form level review. I assume this is to mitigate issues from having all records visible but the similar checking doesn't seem to be in place for the toggle to see all subjects. The toggle doesn't disappear at the form level either so it all feels a bit messy to me.

LDSamson committed:
Use double ampersand instead

LDSamson commented:
> I think generally I'm a bit confused by this PR. In the common events the text wrapping is disabled on form level review. I assume this is to mitigate issues from having all records visible but the similar checking doesn't seem to be in place for the toggle to see all subjects. The toggle doesn't disappear at the form level either so it all feels a bit messy to me.

Hi, thanks for the review. Initially, the idea was to discourage to use it in form level review, but not to disallow it. Do you think it should be completely forbidden?

jthompson-arcus commented:
> > I think generally I'm a bit confused by this PR. In the common events the text wrapping is disabled on form level review. I assume this is to mitigate issues from having all records visible but the similar checking doesn't seem to be in place for the toggle to see all subjects. The toggle doesn't disappear at the form level either so it all feels a bit messy to me.
> 
> Hi, thanks for the review. Initially, the idea was to discourage to use it in form level review, but not to disallow it. Do you think it should be completely forbidden?

If this feature is something desired, I personally would think about turning pagination back on if it is triggered. You are re-rendering the table anyways.

LDSamson commented:
> Enabling text wrap does not appear to be working. For instance checkout subject BEL_04_772 and the Medication tab. There are 15 records. If you "enable text wrap" you can only view 10 of them.

@jthompson-arcus Text wrap works it's just that not all entries are shown indeed which I did not expect.. do you know why this is the case?

I will try to switch to pagination again and see if that works.

LDSamson committed:
Merge branch 'dev' into ls_224_add_table_text_wrap_toggle

# Conflicts:
#	tests/testthat/_snaps/app_feature_01/app-feature-1-003_.png
#	tests/testthat/_snaps/app_feature_03/app-feature-3-001.json
#	tests/testthat/_snaps/app_feature_03/app-feature-3-002.json

LDSamson committed:
Bump version

LDSamson committed:
Update snaps again

LDSamson committed:
Merge branch 'dev' into ls_224_add_table_text_wrap_toggle

# Conflicts:
#	DESCRIPTION
#	NEWS.md
#	R/mod_review_form_tbl.R
#	R/mod_study_forms.R
#	inst/golem-config.yml
#	tests/testthat/test-app_feature_03.R

LDSamson committed:
Enable pagination again conditionally

LDSamson committed:
Update snapshot

LDSamson commented:
> The toggle doesn't disappear at the form level either so it all feels a bit messy to me.

I am a bit confused by this comment.  The toggle should disappear when using form_level review. Is this not the case for you? 
I updated the PR, pagination is now again enabled when text_wrap is enabled. Let me know what you think! 

text wrap can indeed still be enabled when pressing show_all_participants but not when using form_level review, I thought for the latter it is probably not needed.

LDSamson committed:
Use double ampersand for consistency

jthompson-arcus reviewed (changes requested):
I still do not understand why we would not want the end user to be able to enable wrapping when reviewing by form. The word wrap causes the table to display paginated, so we should allow that view for form level reviews as well.

Also, perhaps padding between the AE tables should be considered.

LDSamson committed:
Merge branch 'dev' into ls_224_add_table_text_wrap_toggle

LDSamson committed:
Delete tooltip warning

LDSamson committed:
Enable text wrap toggle in form review mode

LDSamson committed:
Add some padding at the bottom of the SAE table

LDSamson committed:
Make sure the toggle works in common_forms for form-level review

LDSamson committed:
Add text wrap option in form level review for study forms

LDSamson committed:
Toggle text wrap to FALSE when switching to view data from all subjects

LDSamson commented:
Text wrap is now consistent between form-level review and just pressing the `show all participants` button. It can be used with both. 

It switches to no text wrapping when clicking on 'show all participants'. This is because in this mode, it is probably easier to compare lots of data if the data is on one row. But you can still switch on text wrap if desired.

I also added extra padding under the SAE table as suggested and removed the tooltip warning since it seems like text wrap is still okay for large datasets when pagination is enabled (which is the case here)

jthompson-arcus reviewed (changes requested):
I know this is turning into the PR of a thousand change requests but...I found a bug.

If you select rows to review and then toggle "enable text wrapping", the rows will still be cached and you can hit the "save" button but the rows in the table do not show being selected.

LDSamson committed:
Merge branch 'dev' into ls_224_add_table_text_wrap_toggle

LDSamson committed:
Bump version

LDSamson committed:
Merge branch 'dev' into ls_224_add_table_text_wrap_toggle

# Conflicts:
#	DESCRIPTION
#	NEWS.md
#	inst/golem-config.yml
#	tests/testthat/test-app_feature_03.R

LDSamson committed:
Fix pending review records not updating after toggling table wrap.

LDSamson committed:
Add test to ensure pending reviews are unaffected from text wrap toggle

LDSamson commented:
> I know this is turning into the PR of a thousand change requests but...I found a bug.
> 
> If you select rows to review and then toggle "enable text wrapping", the rows will still be cached and you can hit the "save" button but the rows in the table do not show being selected.

@jthompson-arcus thanks for spotting this. Now resolved in this commit: 2c0a2e14f1779abd0ea118d3d35d8886d8ad67da.
I also added a test to ensure that it is resolved: 2c0a2e14f1779abd0ea118d3d35d8886d8ad67da

LDSamson commented on R/mod_review_form_tbl.R:
```diff
@@ -149,9 +154,9 @@ mod_review_form_tbl_server <- function(
       table_data(df)
     })
     
-    # Any time the data in the form table is updated, "show all" is toggled,
-    # or the subject being viewed is changed, the server data for the datatable
-    # needs to be updated
+    # Triggers when server data needs to be updated. Also triggers for each 
+    # change in pending review records (e.g. a checkbox in column `Reviewed` 
+    # is toggled on or off).
```
@jthompson-arcus changed the description here and in lines 173-174 a bit to capture the 'why' because it always take me some time to familiarize with the logic here after a while. If you don't like it or have suggestions for better description please let me know.

jthompson-arcus reviewed (approved):
Looks good to me now. Thanks for the updates.
